src/safeposix/vmmap.rs

```diff
@@ -0,0 +1,587 @@
+        &mut self,
+        page_num: u32,
+        npages: u32,
+        prot: i32,
+        maxprot: i32,
+        flags: i32,
+        backing: MemoryBackingType,
+        file_offset: i64,
+        file_size: i64,
+        cage_id: u64,
+    ) -> Result<(), io::Error>;
+
+    fn change_prot(&mut self, page_num: u32, npages: u32, new_prot: i32);
+
+    fn remove_entry(&mut self, page_num: u32, npages: u32) -> Result<(), io::Error>;
+
+    fn check_existing_mapping(&self, page_num: u32, npages: u32, prot: i32) -> bool;
+
+    fn check_addr_mapping(&mut self, page_num: u32, npages: u32, prot: i32) -> Option<u32>;
+
+    fn find_page(&self, page_num: u32) -> Option<&VmmapEntry>;
+
+    fn find_page_mut(&mut self, page_num: u32) -> Option<&mut VmmapEntry>;
+
+    fn find_page_iter(
+        &self,
+        page_num: u32,
+    ) -> impl DoubleEndedIterator<Item = (&Interval<u32>, &VmmapEntry)>;
+
+    fn find_page_iter_mut(
+        &mut self,
+        page_num: u32,
+    ) -> impl DoubleEndedIterator<Item = (&Interval<u32>, &mut VmmapEntry)>;
+
+    fn first_entry(&self) -> Option<(&Interval<u32>, &VmmapEntry)>;
+
+    fn last_entry(&self) -> Option<(&Interval<u32>, &VmmapEntry)>;
+
+    fn double_ended_iter(&self) -> impl DoubleEndedIterator<Item = (&Interval<u32>, &VmmapEntry)>;
+
+    fn double_ended_iter_mut(
+        &mut self,
+    ) -> impl DoubleEndedIterator<Item = (&Interval<u32>, &mut VmmapEntry)>;
+
+    fn find_space(&self, npages: u32) -> Option<Interval<u32>>;
+
+    fn find_space_above_hint(&self, npages: u32, hint: u32) -> Option<Interval<u32>>;
+
+    fn find_map_space(&self, num_pages: u32, pages_per_map: u32) -> Option<Interval<u32>>;
+
+    fn find_map_space_with_hint(
+        &self,
+        num_pages: u32,
+        pages_per_map: u32,
+        hint: u32,
+    ) -> Option<Interval<u32>>;
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct Vmmap {
+    pub entries: NoditMap<u32, Interval<u32>, VmmapEntry>, // Keyed by `page_num`
+    pub cached_entry: Option<VmmapEntry>,                  // TODO: is this still needed?
+                                                           // Use Option for safety
+}
+
+#[allow(dead_code)]
+impl Vmmap {
+    pub fn new() -> Self {
+        Vmmap {
+            entries: NoditMap::new(),
+            cached_entry: None,
+        }
+    }
+
+    fn round_page_num_up_to_map_multiple(&self, npages: u32, pages_per_map: u32) -> u32 {
+        (npages + pages_per_map - 1) & !(pages_per_map - 1)
+    }
+
+    fn trunc_page_num_down_to_map_multiple(&self, npages: u32, pages_per_map: u32) -> u32 {
+        npages & !(pages_per_map - 1)
+    }
+
+    fn visit() {}
+
+    fn debug() {}
+}
+
+impl VmmapOps for Vmmap {
+    fn add_entry(&mut self, vmmap_entry_ref: VmmapEntry) {
+        let _ = self.entries.insert_strict(
+            // pages x to y, y included
+            ie(
+                vmmap_entry_ref.page_num,
+                vmmap_entry_ref.page_num + vmmap_entry_ref.npages,
```

Comment:
passing an vmmap entry with npages equals to 0 would cause panic here. If npages is not supposed to be 0, then there should be a check at vmmap entry creation or here to prevent from this.